rasa_core_sdk/forms.py

```diff
@@ -72,39 +72,41 @@ def extract(self,
-            for slot_mapping in slot_mappings:
-                if (not isinstance(slot_mapping, dict) or
-                        slot_mapping.get("type") is None):
-                    raise TypeError("Provided incompatible slot_mapping")
-
-                mapping_intent = slot_mapping.get("intent")
-                intent = tracker.latest_message.get("intent",
-                                                    {}).get("name")
-                if mapping_intent is None or mapping_intent == intent:
-                    mapping_type = slot_mapping["type"]
-
-                    if mapping_type == "from_entity":
-                        entity_value = next(tracker.get_latest_entity_values(
-                                slot_mapping.get("entity")), None)
-                        if entity_value is not None:
-                            return [SlotSet(slot_to_fill, entity_value)]
-
-                    elif mapping_type == "from_intent":
-                        return [SlotSet(slot_to_fill,
-                                        slot_mapping.get("value"))]
-
-                    elif mapping_type == "from_text":
-                        return [SlotSet(slot_to_fill,
-                                        tracker.latest_message.get("text"))]
-
-                    else:
-                        raise ValueError('Provided slot_mapping["type"] '
-                                         'is not supported')
+        if not slot_mappings:
+            slot_mappings = self.from_entity(slot_to_fill)
+
+        if not isinstance(slot_mappings, list):
+            slot_mappings = [slot_mappings]
+
+        for slot_mapping in slot_mappings:
+            if (not isinstance(slot_mapping, dict) or
+                    slot_mapping.get("type") is None):
+                raise TypeError("Provided incompatible slot_mapping")
+
+            mapping_intent = slot_mapping.get("intent")
+            intent = tracker.latest_message.get("intent",
+                                                {}).get("name")
+            if mapping_intent is None or mapping_intent == intent:
+                mapping_type = slot_mapping["type"]
+
+                if mapping_type == "from_entity":
+                    entity_value = next(tracker.get_latest_entity_values(
+                            slot_mapping.get("entity")), None)
+                    if entity_value is not None:
+                        return [SlotSet(slot_to_fill, entity_value)]
+
+                elif mapping_type == "from_intent":
+                    return [SlotSet(slot_to_fill,
+                                    slot_mapping.get("value"))]
+
+                elif mapping_type == "from_text":
+                    return [SlotSet(slot_to_fill,
+                                    tracker.latest_message.get("text"))]
+
+                else:
+                    raise NotImplementedError(
+                            'Provided slot_mapping["type"] '
+                            'is not supported')
```

Comment:
I put `NotImplementedError` because, if a user provided some `type` that `extract` do not know, it kind of means that handling of this `type` is not implemented. But I'm not sure if it is correct logic

Reply:
I mean, that usually means the function isn't implemented though right?